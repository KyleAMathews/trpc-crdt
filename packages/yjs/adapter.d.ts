import * as Y from 'yjs';
import { AnyRouter } from '@trpc/server';

interface OnErrorParams {
    error: Error;
    path: string;
    type: string;
    ctx: any;
    input: any;
}
interface AdapterArgs {
    doc: Y.Doc;
    appRouter: AnyRouter;
    context: any;
    onError?: (params: OnErrorParams) => void;
}
declare function adapter({ doc, appRouter, context, onError }: AdapterArgs): void;

export { adapter };
