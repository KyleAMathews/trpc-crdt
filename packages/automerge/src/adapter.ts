import { AnyRouter, callProcedure, getTRPCErrorFromUnknown } from "@trpc/server"
import { getErrorShape } from "@trpc/server/shared"
import { AutomergeUrl, DocHandle, Repo } from "@automerge/automerge-repo"

export type CallState = `WAITING` | `DONE` | `ERROR`

export interface Call {
  path: string
  input: unknown
  type: string
  state: CallState
  started: string
  clientId: string
  elapsedMs: number
  createdAt: string
  response: unknown
}

export interface CallQueue {
  queue: AutomergeUrl[]
}

export interface User {
  id: number
  name: string
}

export interface Users {
  users: User[]
}

export interface OnErrorParams {
  error: Error
  path: string
  type: string
  ctx: unknown
  input: unknown
}

export interface AdapterArgs<T> {
  repo: Repo
  queueHandle: DocHandle<CallQueue>

  appRouter: AnyRouter
  ctx: T
  onError?: (params: OnErrorParams) => void
}

export function adapter<T>({
  repo,
  queueHandle,
  appRouter,
  ctx,
  onError,
}: AdapterArgs<T>) {
  queueHandle.on(`change`, ({ handle }) => {
    handle.change(async (doc: CallQueue) => {
      let nextCall
      while ((nextCall = doc.queue.pop())) {
        const callHandle = repo.find(nextCall)
        const { state, path, input, type } = await callHandle.doc()

        if (state == `WAITING`) {
          try {
            const result = await callProcedure({
              procedures: appRouter._def.procedures,
              path,
              rawInput: input,
              type,
              ctx,
            })

            callHandle.change((d: Call) => {
              d.state = `DONE`
              d.response = result
            })
          } catch (cause) {
            const error = getTRPCErrorFromUnknown(cause)

            onError?.({
              error,
              type,
              path,
              input,
              ctx,
            })

            const errorShape = getErrorShape({
              config: appRouter._def._config,
              error,
              type,
              path,
              input,
              ctx,
            })

            console.log({ error, errorShape })

            callHandle.change((d: Call) => {
              d.state = `ERROR`
              d.response = { error: errorShape }
            })
          }
        }
      }
    })
  })
}
