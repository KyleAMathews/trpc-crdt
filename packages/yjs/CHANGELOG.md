# trpc-yjs

## 0.0.2

### Patch Changes

- 14120dd: Initial setup of repo and release of trpc-yjs
- c72585d: Fixes and API cleanups
