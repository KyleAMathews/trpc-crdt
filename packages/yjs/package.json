{
  "name": "trpc-yjs",
  "description": "Call server functions natively in your Yjs app with tRPC",
  "version": "0.0.9",
  "author": "Kyle Mathews <mathews.kyle@gmail.com>",
  "devDependencies": {
    "@trpc/client": "^10.38.5",
    "@trpc/server": "^10.38.5",
    "yjs": "^13.6.8"
  },
  "exports": {
    "./adapter": "./dist/adapter.js",
    "./link": "./dist/link.js"
  },
  "files": [
    "dist",
    "*.d.ts"
  ],
  "keywords": [],
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "peerDependencies": {
    "yjs": "*",
    "@trpc/client": "*",
    "@trpc/server": "*"
  },
  "scripts": {
    "build": "npm run clean && tsup && npm run post-build",
    "check": "tsc",
    "clean": "shx rm -rf dist *.d.ts",
    "post-build": "shx mv dist/*.d.ts* .",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "tsup": {
    "entry": [
      "src/adapter.ts",
      "src/link.ts"
    ],
    "format": [
      "esm",
      "cjs"
    ],
    "dts": true
  }
}
