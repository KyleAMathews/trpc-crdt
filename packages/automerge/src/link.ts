import type { AnyRouter } from "@trpc/server"
import { DocHandle, Repo } from "@automerge/automerge-repo"
import { TRPCClientError, TRPCLink } from "@trpc/client"
import { observable } from "@trpc/server/observable"
import { Call, CallQueue } from "./adapter"

export const link = <TRouter extends AnyRouter>({
  repo,
  queueHandle,
}: {
  repo: Repo
  queueHandle: DocHandle<CallQueue>
}): TRPCLink<TRouter> => {
  return () =>
    ({ op }) =>
      observable((observer) => {
        // Create a job and populate it with some metadata
        const callHandle = repo.create<Call>()
        callHandle.change((d: Call) =>
          Object.assign(d, {
            ...op,
            response: {},
            createdAt: new Date().toJSON(),
            state: `WAITING`,
          })
        )

        queueHandle.change((d: CallQueue) => {
          d.queue.push(callHandle.url)
        })

        callHandle.on(`change`, ({ doc }: { doc: Call }) => {
          const { state, response } = doc
          if (state === `WAITING`) {
            // no action required here
            return
          }

          // Not waiting, so we must be DONE or ERROR
          callHandle.change((d: Call) => {
            d.elapsedMs = new Date().getTime() - new Date(d.createdAt).getTime()
          })

          if (state === `ERROR`) {
            observer.error(TRPCClientError.from(response.error))
          } else {
            observer.next({
              result: {
                type: `data`,
                data: response,
              },
            })
            observer.complete()
          }
        })
      })
}
