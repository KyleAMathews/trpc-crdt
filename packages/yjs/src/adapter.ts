import * as Y from "yjs"
import { AnyRouter, callProcedure, getTRPCErrorFromUnknown } from "@trpc/server"
import { getErrorShape } from "@trpc/server/shared"

interface OnErrorParams {
  error: Error
  path: string
  type: string
  ctx: any
  input: any
}

// Define the expected types for the arguments
interface AdapterArgs {
  appRouter: AnyRouter
  // eslint-disable-next-line
  // TODO fix type so other keys are allowed.
  context: { doc: Y.Doc }
  onError?: (params: OnErrorParams) => void
}

export function adapter({ appRouter, context, onError }: AdapterArgs) {
  const { doc } = context
  const requests = doc.getArray(`trpc-calls`)
  requests.observe(async (event: Y.YArrayEvent<any>) => {
    const { insert } = (event?.changes?.delta as any[]).find((item) => {
      return `insert` in item
    }) || { insert: [] }
    insert.forEach(async (state: any) => {
      // Backwards compatability for older calls that were plain objects.
      if (!state.get) {
        return
      }
      if (state.get(`state`) === `WAITING`) {
        const transactionFns: any[] = []
        const transact = (fn: () => void) => {
          transactionFns.push(fn)
        }
        try {
          await callProcedure({
            procedures: appRouter._def.procedures,
            path: state.get(`path`),
            rawInput: state.get(`input`),
            type: state.get(`type`),
            ctx: { ...context, transact, response: state.get(`response`) },
          })
          doc.transact(() => {
            transactionFns.forEach((fn) => {
              fn()
            })
            state.set(`state`, `DONE`)
          })
        } catch (cause) {
          const error = getTRPCErrorFromUnknown(cause)
          const errorShape = getErrorShape({
            config: appRouter._def._config,
            error,
            type: state.get(`type`),
            path: state.get(`path`),
            input: state.get(`input`),
            ctx: context,
          })

          onError?.({
            error,
            path: state.get(`path`),
            type: state.get(`type`),
            ctx: context,
            input: state.get(`input`),
          })

          doc.transact(() => {
            state.set(`state`, `ERROR`)
            state.get(`response`).set(`error`, { error: errorShape })
          })
        }
      }
    })
  })
}
