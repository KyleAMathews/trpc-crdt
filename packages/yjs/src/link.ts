import type { AnyRouter } from "@trpc/server"
import { Doc, YMapEvent, Map } from "yjs"
import { v4 as uuidv4 } from "uuid"
import { TRPCClientError, TRPCLink } from "@trpc/client"
import { observable } from "@trpc/server/observable"

export const link = <TRouter extends AnyRouter>({
  doc,
}: {
  doc: Doc
}): TRPCLink<TRouter> => {
  return () =>
    ({ op }) =>
      observable((observer) => {
        const calls = doc.getArray(`trpc-calls`)
        const requestId = uuidv4()
        const requestMap = new Map()

        // The observe function to listen to the response
        // from the server.
        function observe(event: YMapEvent<any>) {
          const state = event.target
          if (state.get(`done`) && state.get(`id`) === requestId) {
            requestMap.unobserve(observe)
            requestMap.set(
              `elapsedMs`,
              new Date().getTime() -
                new Date(requestMap.get(`createdAt`)).getTime()
            )
            if (state.get(`error`)) {
              observer.error(TRPCClientError.from(state.get(`response`)))
            } else {
              observer.next({
                result: {
                  type: `data`,
                  data: state.get(`response`),
                },
              })
            }
            observer.complete()
          }
        }

        // Create the request map on the trpc-calls Y.js Array.
        // This will get replicated to the server.
        const { path, input, type } = op
        console.log({ path, input, type })

        doc.transact(() => {
          requestMap.set(`path`, path)
          requestMap.set(`input`, input)
          requestMap.set(`type`, type)
          requestMap.set(`id`, requestId)
          requestMap.set(`done`, false)
          requestMap.set(`createdAt`, new Date().toJSON())
          requestMap.set(`clientId`, doc.clientID)
          requestMap.observe(observe)
          calls.push([requestMap])
        })
      })
}
