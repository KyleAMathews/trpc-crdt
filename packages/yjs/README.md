# trpc-yjs

[tRPC](https://trpc.io/) integration for [Yjs](https://yjs.dev/): Yjs-native RPC calls

For when you're building with Yjs but you want to run code on a server.

See the [monorepo README for a longer writeup on why this library exists](https://github.com/KyleAMathews/trpc-crdt).

## Install

`npm install trpc-yjs`

## Example app

https://trpc-yjs.fly.dev/

## Usage

#### Browser

```ts
import * as Y from "yjs"
import { createTRPCProxyClient } from "@trpc/client"
import { link } from "trpc-yjs/link"

// Doc needs replicated via a server e.g. with y-websocket.
const doc = new Y.Doc()
const trpc = createTRPCProxyClient<AppRouter>({
  links: [
    loggerLink(),
    link({
      doc,
    }),
  ],
})

// Tell server to create a new user.
await trpc.userCreate.mutate({id: `1`, name: `Kyle Mathews`})

// The new user, written by the server, is now available at
doc.getMap(`users`).get(`1`)
```

#### Server

```ts
import { adapter } from "trpc-yjs/adapter"
import { initTRPC, TRPCError } from "@trpc/server"
import { z } from "zod"

const t = initTRPC.create()
const router = t.router
const publicProcedure = t.procedure

const appRouter = router({
  userCreate: publicProcedure
    .input(z.object({ name: z.string(), id: z.string() }))
    .mutation(async (opts) => {
      const {
        input,
        ctx: { users },
      } = opts
      const user = { ..input }

      // Set new user on the Y.Map users.
      users.set(user.id, user)

      return `ok`
    })
})

// Get replicated Yjs doc to listen for new tRPC calls.
const doc = getYDoc(`doc`)

// Setup trpc-yjs adapter.
adapter({ appRouter, context: { doc, users: doc.getMap(`users`) } })
```

## Considerations

This library serializes RPC calls between clients and server through `trpc-call` objects in a Yjs array.

So to use, you need a Node.js tRPC server that's listening for updates to the shared doc to respond to users. If using y-websocket, this can be the same Node.js process running websockets. If you're using a hosted Yjs service, the Node.js process can run on any Node.js host and connect to the shared server.

There's a variety of ways you can configure trpc-yjs. You can use a shared doc for all users, a doc / user. If your app has a high volume of server calls, you might want to rotate docs e.g. daily or even hourly to avoid too much data accumulating.

This package is new so feel free to open issues if you have questions or encounter bugs.
