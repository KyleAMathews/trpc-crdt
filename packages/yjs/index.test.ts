import { describe, it, expect } from "vitest"
import { initTRPC, TRPCError } from "@trpc/server"
import { z } from "zod"
import * as Y from "yjs"
import { adapter } from "./src/adapter"
import { link } from "./src/link"
import { createTRPCProxyClient, loggerLink, httpBatchLink } from "@trpc/client"
/**
 * Initialization of tRPC backend
 * Should be done only once per backend!
 */
const t = initTRPC.create()
/**
 * Export reusable router and procedure helpers
 * that can be used throughout the router
 */
const router = t.router
const publicProcedure = t.procedure

const appRouter = router({
  userCreate: publicProcedure
    .input(z.object({ name: z.string(), optionalDelay: z.number().optional() }))
    .mutation(async (opts) => {
      const {
        input,
        ctx: { users },
      } = opts
      const user = { id: String(users.length + 1), ...input }

      if (input.optionalDelay) {
        await new Promise((resolve) => setTimeout(resolve, input.optionalDelay))
      }

      if (input.name === `BAD_NAME`) {
        throw new TRPCError({
          code: `CONFLICT`,
          message: `This name isn't one I like to allow`,
        })
      }

      return {
        transact: () => {
          users.push([user])
        },
        response: user,
      }
    }),
  userUpdateName: publicProcedure
    .input(z.object({ id: z.string(), name: z.string() }))
    .mutation(async (opts) => {
      const {
        input,
        ctx: { users },
      } = opts
      let user
      let id
      users.forEach((u, i) => {
        if (u.id === input.id) {
          user = u
          id = i
        }
      })
      const newUser = { ...user, name: input.name }
      return {
        mutations: () => {
          users.delete(id, 1)
          users.insert(id, [newUser])
        },
        response: newUser,
      }
    }),
})

type AppRouter = typeof appRouter

function initClient() {
  const serverDoc = new Y.Doc()
  const clientDoc = new Y.Doc()

  serverDoc.on(`update`, (update) => {
    Y.applyUpdate(clientDoc, update)
  })

  clientDoc.on(`update`, (update) => {
    Y.applyUpdate(serverDoc, update)
  })

  // YJS database
  const serverUsers = serverDoc.getArray(`users`)

  // Start adapter
  adapter({ doc: serverDoc, appRouter, context: { users: serverUsers } })

  // Create client.
  const trpc = createTRPCProxyClient<AppRouter>({
    links: [
      link({
        doc: clientDoc,
      }),
    ],
  })

  return { doc: clientDoc, trpc }
}

describe(`yjs`, () => {
  describe(`basic calls`, () => {
    const { trpc } = initClient()
    it(`create a user`, async () => {
      const newUser = await trpc.userCreate.mutate({ name: `foo` })
      expect(newUser.name).toEqual(`foo`)
    })
    it(`updateName`, async () => {
      const user = await trpc.userUpdateName.mutate({ id: `1`, name: `foo2` })
      expect(user.name).toEqual(`foo2`)
    })
  })
  describe(`batched calls`, () => {
    const { doc, trpc } = initClient()
    it(`handles batched calls`, async () => {
      let promise1
      let promise2
      doc.transact(() => {
        promise1 = trpc.userCreate.mutate({ name: `foo1` })
        promise2 = trpc.userCreate.mutate({ name: `foo2` })
      })

      await Promise.all([promise1, promise2])

      let promise3
      let promise4

      doc.transact(() => {
        promise3 = trpc.userCreate.mutate({ name: `foo3` })
        promise4 = trpc.userCreate.mutate({ name: `foo4` })
      })

      await Promise.all([promise3, promise4])

      await trpc.userCreate.mutate({ name: `foo5` })

      const users = doc.getArray(`users`).toJSON()

      expect(users).toHaveLength(5)
    })
  })
  describe(`out-of-order calls`, () => {
    const { trpc } = initClient()
    it(`handles out-of-order calls`, async () => {
      const user1Promise = trpc.userCreate.mutate({
        name: `foo1`,
        optionalDelay: 10,
      })
      const user2Promise = trpc.userCreate.mutate({ name: `foo2` })

      const [user1, user2] = await Promise.all([user1Promise, user2Promise])

      expect(user1.name).toEqual(`foo1`)
      expect(user2.name).toEqual(`foo2`)
    })
  })
  describe(`handle errors`, () => {
    const { trpc } = initClient()
    it(`input errors`, async () => {
      await expect(() =>
        trpc.userCreate.mutate({ name: 1 })
      ).rejects.toThrowError(`invalid_type`)
    })
    it(`router thrown errors`, async () => {
      await expect(() =>
        trpc.userCreate.mutate({ name: `BAD_NAME` })
      ).rejects.toThrowError(`This name isn't one I like to allow`)
    })
  })
})
