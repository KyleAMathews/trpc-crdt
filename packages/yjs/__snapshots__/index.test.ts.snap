// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`yjs > basic calls > create a user 1`] = `
[
  {
    "id": "1",
    "name": "foo",
  },
]
`;

exports[`yjs > basic calls > create a user 2`] = `
[
  {
    "id": "1",
    "name": "foo2",
  },
]
`;

exports[`yjs > basic calls > updateName 1`] = `
[
  {
    "id": "1",
    "name": "foo2",
  },
]
`;
